wire-runtime-swift/src/main/swift/ProtoCodable/ProtoDecoder.swift

```diff
@@ -131,11 +131,132 @@ public final class ProtoDecoder {
         }
     }
 
+    // MARK: - Internal Methods
+
+    /** Decode a tagged `ProtoDecodable` field from raw data */
+    internal func decode<T: ProtoDecodable>(_ type: T.Type, from data: Data, withTag tag: UInt32) throws -> T {
+        try decodeWithReader(from: data, emptyValue: nil) { reader in
+            try reader.decode(type, withTag: tag)
+        }
+    }
+
+    /** Decode a tagged `ProtoEnum` field from raw data */
+    internal func decode<T: ProtoEnum>(_ type: T.Type, from data: Data, withTag tag: UInt32) throws -> T where T: RawRepresentable<Int32> {
+        try decodeWithReader(from: data, emptyValue: nil) { reader in
+            try reader.decode(type, withTag: tag)
+        }
+    }
+
+    /** Decode a tagged `int32`, `sfixed32`, or `sint32` field from raw data */
+    internal func decode(_ type: Int32.Type, from data: Data, encoding: ProtoIntEncoding, withTag tag: UInt32) throws -> Int32 {
```

Comment:
Some methods here can probably be reduced by using generics, specifically `ProtoIntCodable`. However, upon more spelunking we saw that this is somewhat intentional for performance reasons to avoid runtime look up of which method to invoke.